### PR TITLE
Consolidate Select component

### PR DESCRIPTION
**Pull request by tlrobinson**
Our `Select` widget is actually two different components the old `LegacySelect` which takes an `options` prop (and supports sections), and the newer `BrowserSelect` that tries to match the builtin `<select>` element (plus supports search and multiple selections.

`BrowserSelect` duplicates a fair amount of the logic in `<AccordianList>` (searching, virtualizing the list, grouping, etc) which we use in various other select-like popovers. I think it makes sense to just use `AccordianList`, which also makes `Select` most consistent with other parts of the UI.

old "BrowserSelect"

Screen Shot 2019-12-13 at 5 44 01 PM

old "LegacySelect"
Screen Shot 2019-12-13 at 5 38 05 PM

new Select:

Screen Shot 2019-12-13 at 5 37 55 PM

This is mostly complete but before going any further I'd like to check with @mazameli and @kdoh. Poke around at the examples in http://localhost:3000/_internal/components/select which contain all 3 (browser, legacy, "new") with various features (search, multiple, sections) enabled

### CODE REVIEW

**Comment by kdoh** [comment]
Overall looking pretty nice. I pushed one small change to tweak the description text a bit to make it a little easier to read.

Process wise is the idea that once we like the new select we'd replace `LegacySelect` with it? And then remove legacy select and / or `BrowserSelect` so we're just down to one?

One thing that I'm noticing (might not be new but focusing on the select is making it hard to unsee) is how the select kinda jumps on first open when it's figuring out if it can render below or has to do it above. In a similar vein, it'd be nice if we could ease the jump in height when the results are filtered down during search. We should also see if we can come up with a nicer no search results state than we have now.

**Comment by kdoh** [comment]
Screen Shot 2019-12-16 at 7 47 23 AM

We should even out the top and bottom spacing when there's only one element.

**Comment by tlrobinson** [comment]
> Process wise is the idea that once we like the new select we'd replace LegacySelect with it? And then remove legacy select and / or BrowserSelect so we're just down to one?

Yes, the idea is to just have one `Select` component, the new one.

**Comment by tlrobinson** [comment]
Also, one pattern I'm experimenting with is allowing you to either pass `sections` (like `AccordionList`), `options` (like `LegacySelect` and `Radio`), or `children` (like `BrowserSelect`)

```
<Select options={[{ name: "Foo", value: "foo" }]} />
```

```
<Select>
  <Option value="foo">Foo</Option>
</Select>
```

are equivalent.

If you have multiple sections you can do 

```
<Select>
  <OptionSection name="First">
    <Option value="foo">Foo</Option>
  </OptionSection>
  <OptionSection name="Second">
    <Option value="bar">Bar</Option>
  </OptionSection>
</Select>
```

or 

```
<Select sections={[
  { name: "First", items: [{ name: "Foo", value: "foo" }] }
  { name: "Second", items: [{ name: "Bar", value: "bar" }] }
]}
/>
```

I'm not sure this is a great idea but it's not a ton of code to support it: https://github.com/metabase/metabase/pull/11522/files#diff-5862809497bf3f0723ae2d063c3be95bR65

The JSX version is nice when you have a few hardcoded options, whereas the `options`/`section` is nice when you have a long list of options (also probably better for performance)

It doesn't currently support `optionNameFn`/`optionValueFn` like in `LegacySelect` and `Radio` but it could, which would be even better for performance when there are large numbers of options with name/values that don't match the expected `name`/`value` properties.

**Comment by mazameli** [comment]
One thing I'd love to have with these when there's a search bar is the ability to use the up and down arrow keys and enter to make a selection after typing in the box to narrow the list down since my fingers are already on the keyboard.

Also it seems to me like the difference between a single select and a multi-select is not visually evident at all. Do we need checkboxes for multi-selects, @kdoh?

Also @tlrobinson, in your `new Select` screenshot, I'm seeing the whole popover has a blue focus border around it, which looks strange. This appears to happen whenever clicking on a selection within one of the Selects.

File: frontend/test/metabase/scenarios/admin/databases/add.cy.spec.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    cy.visit("/admin/databases/create");
+
+    typeField("name", "Test db name");
+    typeField("dbname", "test_postgres_db");
+    typeField("user", "uberadmin");
+
+    cy.contains("button", "Save")
+      .should("not.be.disabled")
+      .click();
+    cy.contains("button", "Saving...").should("be.disabled");
```
Comment (paulrosenzweig) [inline frontend/test/metabase/scenarios/admin/databases/add.cy.spec.js]: I think this is technically a race, but it probably won't ever matter. If the save completes before Cypress processes this `contains`, the test will fail.

It doesn't matter as networks calls will probably always lose that race, but it made me wonder if there's a more robust way to write the test. Maybe you'd need to stub the request to ensure that it won't complete too early?

File: frontend/test/metabase/scenarios/admin/datamodel/table.cy.spec.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      return cy
+        .get(`input[value="${name}"]`)
+        .parent()
+        .parent();
+    }
+
+    function testSelect(alias, initialOption, desiredOption) {
+      cy.get(alias)
+        .contains(initialOption)
+        .click({ force: true });
```
Comment (paulrosenzweig) [inline frontend/test/metabase/scenarios/admin/datamodel/table.cy.spec.js]: Why was this force necessary?

**Review by paulrosenzweig — APPROVED** [review]
Ported tests look great! 👏 

I think for snapshot/restore we'll want to just restore before each test suite, but that can be updated as all of this is merged together.

File: frontend/src/metabase/admin/datamodel/components/database/ColumnItem.jsx [hunk 3]
```diff
@@ -140,15 +138,14 @@ export class SpecialTypeAndTargetPicker extends Component {
     selectSeparator?: React$Element<any>,
   };
 
-  onSpecialTypeChange = async ({ id: special_type }) => {
+  handleChangeSpecialType = async ({ target: { value: special_type } }) => {
```
[the remark this answers is not in the document]
Comment (tlrobinson) [reply frontend/src/metabase/admin/datamodel/components/database/ColumnItem.jsx]: Yes, that's idiomatic React. `on*` for props, `handle*` for methods (that are passed to the `on*` props. https://reactjs.org/docs/handling-events.html

**Comment by tlrobinson** [comment]
Depends on #11564 which introduces Cypress test snapshot/restore functionality

File: frontend/test/metabase/scenarios/admin/databases/add.cy.spec.js (hunk 1 above)
Comment (tlrobinson) [reply frontend/test/metabase/scenarios/admin/databases/add.cy.spec.js]: I assume Cypress immediately begins looking for "Saving..." after performing the click. I guess the question is how often it checks. If it's polling at a rate slower than the fastest network request, that could be an issue. If it's listening for DOM mutations it should be fine.

**Pull request merged**